File: longest-palindromic-subsequence/longest-palindromic-subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        string st=s;
        reverse(st.begin(),st.end());
        int size=s.size();
        vector<vector<int>>dp(size+1,vector<int>(size+1,0));
        for(int i=1;i<=size;i++){
           for(int j=1;j<=size;j++){
                if(s[i-1]==st[j-1])
                    dp[i][j]=dp[i-1][j-1]+1;
               else
                   dp[i][j]=max(dp[i-1][j],dp[i][j-1]);
           }
       }
        return dp[size][size];
    }
};

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        st = s[::-1]
        dp = [[-1]*(n+1) for _ in range(n+1)]
        def solve(i,j):
            if i == n or j == n:
                return 0
            if dp[i][j] != -1:
                return dp[i][j]
            if s[i] == st[j]:
                dp[i][j] = solve(i+1,j+1)+1
            else:
                dp[i][j] = max(solve(i+1,j),solve(i,j+1))
            return dp[i][j] 
        return solve(0,0)